site/en/posts/site-refactor-2022.md

```diff
@@ -0,0 +1,92 @@
+- They need to be [responsive](https://web.dev/learn/design/responsive-images/#responsive-images-with-srcset), with the same image being scaled to multiple different sizes.
+- All of those images also need to be for multiple [image formats](https://web.dev/learn/design/picture-element/#image-formats), so we can have the best chance of serving the smallest size image for each user and device.
+- All generated images need to be optimized to reduce overall image size with minimal change in picture quality.
+- Any images whose source is an animated `.gif` needs to be converted to a silent, looping, autoplaying video, which are usually much smaller.
+
+These requirements exist so that we can, to the best of our ability, make sure our users are served only the media that's required for their browsing context, and only the most optimized versions of that media. What this meant in practice, though, was that for every individual image we had on our site, we needed to generate about 15 images in varying formats and potentially support GIF to video encoding. We also needed to rely on developers having a number of additional command-line packages installed that differed across operating systems.
+
+Our first attempt at this, [Eleventy Plugin Local Responsive Images](https://github.com/chromeos/static-site-scaffold-modules/tree/91285a3289358599df58b29190b80e8afd107d59/modules/eleventy-plugin-local-respimg) was able to do this pretty well; it kept an in-memory cache of images it had already optimized so it wouldn't have to do them twice in the same run, was super configurable, and didn't require special syntax to work; just an `img` tag that pointed to a local image, and it magically worked! For videos, too! It was built as an Eleventy transform, which meant that on each build of the site, it would need to do the transforms for all images across all rendered pages. This meant the first build of a watch task, for instance, was very slow, while subsequent builds were faster.
+
+Our second attempt was the previously mentioned Vite Plugin Img and did more or less the same thing, but moved the transforms to PostHTML so they ran only on the page being displayed during development, or only when needed during a production build. It also moved to only rendering the images during a production build, which generally sped up the whole development process. The original goal was to leverage the Vite ecosystem and let other image related plugins handle the optimization, but the images would end up being written too late for common image optimization plugins to pick them up, so we implemented image optimization ourself. This seemed like a good path, until we found a [weird Vite bug](https://github.com/vitejs/vite/issues/5815) where if one of our image optimization dependencies was simply required, not even used, just required, in a Vite config, Sass would fail to load! So, we either needed another solution, or needed to rethink this module in its entirety.
```

Comment:
```suggestion
Our second attempt was the previously mentioned Vite Plugin Img and did more or less the same thing, but moved the transforms to PostHTML so they ran only on the page being displayed during development, or only when needed during a production build. It also moved to only rendering the images during a production build, which generally sped up the whole development process. The original goal was to leverage the Vite ecosystem and let other image related plugins handle the optimization, but the images would end up being written too late for common image optimization plugins to pick them up, so we implemented image optimization ourself. This seemed like a good path, until we found a [weird Vite bug](https://github.com/vitejs/vite/issues/5815) where if one of our image optimization dependencies was simply required, (not even used, just required) in a Vite config, Sass would fail to load! So, we either needed another solution, or needed to rethink this module in its entirety.
```

Reply:
I feel like commentary might be better in parentheses, again just a nit.

Reply:
Resolved manually, with — em dashes —